src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+        }
+
+        // If there are too many files OR (data size is over the threshold and the data set is too wide),
+        // we need a sure shot way of reducing them.
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
+            }
+
+            List<OverlappingSet> overlappingDataSets = getDistinctOverlappingSSTables(maxGarbageData);
+            long windowSzInMs = getWindowSizeInMs();
+
+            logger.debug("{}: too many data files", tableName);
+            // we are here, which means that tombstone.timerange <= 2 * window && tombstone.size() <= 1, and there could be too many data files.
+            // data files could be too wide or many small files.
+            // we would like to merge all those files that are overlapping or are in a single compaction window.
+            return overlappingDataSets.stream()
+                    // group them by compaction window
+                    .collect(Collectors.groupingBy(s -> toWindowMs(s.timeWindowMs.ts, windowSzInMs)))
+                    .values().stream()
+                    // flatten all the sstables present in all of the overlapping sets lying in this compaction window
+                    .map(v -> v.stream()
+                            .flatMap(s -> s.sstables.stream())
+                            .collect(Collectors.toList()))
+                    // get the window with the most sstables
+                    .max(Comparator.comparingInt(List::size))
+                    // limit the count of sstables to maxFileCountForCompaction
+                    .map(l -> limit(getMaxFileCountForCompaction(), l))
+                    // create a compaction task
+                    .map(l -> new SSTablesForCompaction(toSSTableReader(l), false, true));
+        }
+        else
+        {
+            List<SSTableReader> all = new ArrayList<>(toSSTableReader(maxGarbageTombstones.sstables));
+            all.addAll(toSSTableReader(maxGarbageData));
+            return Optional.of(new SSTablesForCompaction(all, false, false));
+        }
+    }
+
+    private Optional<SSTablesForCompaction> getCompactionCandidateBasedOnOverlaps(CategorizedSSTables categorizedSSTables)
+    {
+        int maxFileForCompaction = getMaxFileCountForCompaction();
+        List<SSTableStats> mergeableLatestTombstones = getMaxOverlappingSSTables(categorizedSSTables.latestTombstones);
+        // if latest tombstones are overlapping too much, merge them without splitting.
+        if (mergeableLatestTombstones.size() > 1)
```

Comment:
Why the higher priority for different types of sstable? If the idea is simply to reduce overlap, shouldn't we pick the largest overlappingSet across all categories? (data, expired, latest etc)

Reply:
Its because tombstones files move from latest -> near expiry -> expired category.
all having different characteristcs.
Latest will have new tombstones sstables on every flush and will probably be wide. So we keep merging them to keep number of files in this category low on the first opportunity.

Near expiry is a transition period. near sstables here are added every gcSeconds. It should have low overlap. But if they are wide (as they will be, the result of merging in latest sstables.), split them now so that when they get expired they can be efficiently merged together with the data files.
Then in the last, take care of the all files in the expired tombstones and data files.